LPD-31382 - tooltip does not display correct value for required multiple when a UOM and config exists

wanderlast opened the pull request:
Hi @andrea-ale-sbarra 

https://liferay.atlassian.net/browse/LPD-31382, tied to LPP. This is a fire ticket.

Per Bernard's comment [here](https://liferay.slack.com/archives/G9CMBBA9H/p1720718821538249?thread_ts=1720712439.834709&cid=G9CMBBA9H), the fix essentially tries to find the LCM of the values in the UOM and product config's multiple value.

The LCM solver as written is incredibly messy but is being done in this manner due to the potential case of _both_ the UOM and product config's values being decimals (e.g. 0.5 and 0.3 -- this can be seen in the previously written test case, in fact). Attempting to modulo them lead to the portal hanging while attempting to calculate it (which didn't happen if either of them was greater than 1 -- e.g. 5 and 0.3 worked fine as did 5 and 3). I'm unsure of why JS couldn't handle this (I'm assuming it's due to problems with floating point math), but to circumvent this since this is a fire ticket I basically attempted to find a way to make them into integers. The result is what you see below.

The previously existing tests were edited to account for the corrected tooltip value.

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

wanderlast commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-34.liferay.com/job/test-portal-source-format/3579/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/3579/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:x: ci:test:sf - 0 out of 1 jobs passed in 4 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/18186553726902f517a0c480cc90c839c485a5d1">18186553726902f517a0c480cc90c839c485a5d1</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/wanderlast/liferay-portal/tree/LPD-31382">LPD-31382</a><br/>Branch GIT ID: <a href="https://github.com/wanderlast/liferay-portal/commit/bf62ef8e91e6533e914fb4a2433ef4a84774ec0e">bf62ef8e91e6533e914fb4a2433ef4a84774ec0e</a></p>0 out of 1jobs PASSED<div><h4>1 Failed Jobs:</h4><ul><li><a href="https://test-1-34.liferay.com/job/test-portal-source-format/3579/"><strike>test-portal-source-format</strike></a></li></ul></div><h5>For more details click <a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/3579/jenkins-report.html">here</a>.</h5><pre>     [exec] [4/4] Building fresh packages...
     [exec] Done in 6.51s.
     [exec] 
     [exec] &gt; Task :packageRunCheckFormat
     [exec] yarn run v1.13.0
     [exec] \$ node-scripts format --check
     [exec] 🛫 Running preflight...
     [exec] 📜 Generating tsconfig files...
     [exec] 📜 Validating tsconfig files...
     [exec] 🕵️ Checking modified typescript files...
     [exec] ℹ️ A total of 12 CPUs were detected: launching tsc using 12 workers
     [exec] ℹ️ Going to check 0 modified projects
     [exec] 📝 Running format...
     [exec] 📝 Formatting 2 files
     [exec] Error: Format checked 2 files
     [exec] 2 files have problems
     [exec] 
     [exec]     at format (file:///opt/dev/projects/github/liferay-portal/modules/_node-scripts/format/format.mjs:328:9)
     [exec]     at async main (file:///opt/dev/projects/github/liferay-portal/modules/_node-scripts/format/index.mjs:27:2)
     [exec] error Command failed with exit code 1.
     [exec] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
     [exec] 
     [exec] &gt; Task :packageRunCheckFormat FAILED
     [exec] Gradle build finished at 2024-07-12 05:42:27.500.
     [exec] 
     [exec] 
     [exec] FAILURE: Build failed with an exception.
     [exec] 
     [exec] * What went wrong:
     [exec] Execution failed for task ':packageRunCheckFormat'.
     [exec] &gt; Process 'command '/opt/dev/projects/github/liferay-portal/build/node/bin/node'' finished with non-zero exit value 1
     [exec] 
     [exec] * Try:
     [exec] &gt; Run with --info or --debug option to get more log output.
     [exec] &gt; Get more help at https://help.gradle.org.
     [exec] 
     [exec] * Exception is:
     [exec] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':packageRunCheckFormat'.
     [exec]   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda\$executeIfValid\$1(ExecuteActionsTaskExecuter.java:148)
     [exec]   at org.gradle.internal.Try\$Failure.ifSuccessfulOrElse(Try.java:282)
     [exec]   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:146)
     [exec]   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:134)</pre></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-34.liferay.com/job/test-portal-source-format/3579/">test-portal-source-format#3579</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/3579/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-commerce/liferay-portal/pull/5143">liferay-commerce#5143</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/#/project/34754/routines/34756">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/#/project/34754/routines/34756/build/36402665">[master] ci:test:sf - wanderlast &gt; liferay-commerce - PR#5143 - 2024-07-11[22:38:39]</a><br/></div><div><strong>Testray Build ID:</strong><span></span>36402665<br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-34.liferay.com/job/publish-testray-report/24624/">publish-testray-report#24624</a><br/></div></div>

wanderlast commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-30.liferay.com/job/test-portal-source-format/5894/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-30.liferay.com/userContent/jobs/test-portal-source-format/builds/5894/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 4 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/18186553726902f517a0c480cc90c839c485a5d1">18186553726902f517a0c480cc90c839c485a5d1</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/wanderlast/liferay-portal/tree/LPD-31382">LPD-31382</a><br/>Branch GIT ID: <a href="https://github.com/wanderlast/liferay-portal/commit/1650f14c81cb00cbc669a6e6c431c12ff804117e">1650f14c81cb00cbc669a6e6c431c12ff804117e</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-30.liferay.com/job/test-portal-source-format/5894/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-30.liferay.com/userContent/jobs/test-portal-source-format/builds/5894/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-30.liferay.com/job/test-portal-source-format/5894/">test-portal-source-format#5894</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-30.liferay.com/userContent/jobs/test-portal-source-format/builds/5894/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-commerce/liferay-portal/pull/5143">liferay-commerce#5143</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/#/project/34754/routines/34756">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/#/project/34754/routines/34756/build/36409413">[master] ci:test:sf - wanderlast &gt; liferay-commerce - PR#5143 - 2024-07-11[23:31:23]</a><br/></div><div><strong>Testray Build ID:</strong><span></span>36409413<br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-30.liferay.com/job/publish-testray-report/22997/">publish-testray-report#22997</a><br/></div></div>

riccardo-alberti commented on modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/quantities.js:
```diff
@@ -25,19 +25,40 @@ export function getMinQuantity(
-		multiple += incrementalOrderQuantity;
+
+	while (multiple % itemMultiple !== 0) {
+		multiple += unitOfMeasureMultiple;
```
@wanderlast is it not easier to use something like this?
function findLCM(a, b) { 
    let lar = Math.max(a, b); 
    let small = Math.min(a, b); 
    for (i = lar; ; i += lar) { 
        if (i % small == 0) 
            return i; 
    } 
} 
will this not solve for all cases?

wanderlast replied:
Like I mentioned, %/modulus when used with two decimal values below 1 causes the portal to infinite loop/lag out due to it technically never reaching a valid value when evaluating i % small. You can see an example of this in the screenshot below where I used the function you provided:

![Screenshot 2024-07-12 115653](https://github.com/user-attachments/assets/951232e3-cad1-40ae-a361-08bb269427a8)

Note the variable values on the left.

wanderlast commented:
Changes made:
- I realized that there are some decimal values that can exist that still fall prey to the floating point math errors that are greater than 1, so I've removed that. I also improved how the precision is calculated so no precision should be lost.
- I changed the LCM logic to include the min/max calculation and the for loop per your suggestion.

I unfortunately still have not found a way for this to work without the adjustments to float values to make them integers.

riccardo-alberti reviewed (approved):
for the lpp it's ok like this. For master we might want to find an easier way to calculate this. thanks @wanderlast
